### PR TITLE
Pac4j authentication handler should test for AJAX request before redirection.

### PR DESCRIPTION
**Pull request by jeff-blaisdell**
Initial attempt to fix https://github.com/ratpack/ratpack/issues/538

### CODE REVIEW

File: ratpack-pac4j/src/main/java/ratpack/pac4j/internal/Pac4jAuthenticationHandler.java [hunk 1]
```diff
@@ -76,4 +78,9 @@ private void initiateAuthentication(final Context context) {
       }
     }).then(ignored -> webContext.sendResponse());
   }
+
+  private boolean isAjaxRequest(Request request) {
+    Headers headers = request.getHeaders();
+    return "XMLHttpRequest".equals(headers.get(HttpHeaderConstants.X_REQUESTED_WITH));
```
Comment (davidmc24) [inline ratpack-pac4j/src/main/java/ratpack/pac4j/internal/Pac4jAuthenticationHandler.java]: Might be worthwhile to move `"XMLHttpRequest"` into `HttpHeaderConstants` as well, actually.

Comment (johnrengelman) [inline ratpack-pac4j/src/main/java/ratpack/pac4j/internal/Pac4jAuthenticationHandler.java]: Probably is worthwhile to have this method defined on the `Request` API itself, don't you think? I can image there are a number of place where you'd like to know if a request is ajax or not.

Comment (davidmc24) [inline ratpack-pac4j/src/main/java/ratpack/pac4j/internal/Pac4jAuthenticationHandler.java]: Yeah, you're probably right.  I'm always hesitant to make hard use of non-standard HTTP headers, but this one is probably a case where the value is worth it.

Comment (jeff-blaisdell) [inline ratpack-pac4j/src/main/java/ratpack/pac4j/internal/Pac4jAuthenticationHandler.java]: Good points.  I'll update.

Comment (jeff-blaisdell) [inline ratpack-pac4j/src/main/java/ratpack/pac4j/internal/Pac4jAuthenticationHandler.java]: Above changes have been made.

File: ratpack-core/src/main/java/ratpack/http/Request.java [hunk 2]
```diff
@@ -130,6 +130,12 @@
   HostAndPort getLocalAddress();
 
   /**
+   * A flag representing whether or not the request originated via AJAX.
+   * @return A flag representing whether or not the request originated via AJAX.
+   */
+  Boolean isAjaxRequest();
```
Comment (davidmc24) [inline ratpack-core/src/main/java/ratpack/http/Request.java]: If it's not meaningful for this to be `null`, might want to change this to return `boolean`.

File: ratpack-core/src/main/java/ratpack/http/internal/DefaultRequest.java [hunk 3]
```diff
@@ -185,6 +185,10 @@ public String oneCookie(String name) {
     }
   }
 
+  public Boolean isAjaxRequest() {
+    return HttpHeaderConstants.XML_HTTP_REQUEST.equals(headers.get(HttpHeaderConstants.X_REQUESTED_WITH));
```
Comment (davidmc24) [inline ratpack-core/src/main/java/ratpack/http/internal/DefaultRequest.java]: Should this match be case-sensitive, or case-insensitive?

**Comment by davidmc24** [comment]
Pretty close.  I think with the next round of changes this will be ready to merge.  Please add yourself to the list of contributors in `/ratpack-manual/src/content/chapters/99-about-the-project.md`.

**Comment by jeff-blaisdell** [comment]
Made changes.

File: ratpack-core/src/main/java/ratpack/http/internal/DefaultRequest.java (hunk 3 above)
Comment (jeff-blaisdell) [inline ratpack-core/src/main/java/ratpack/http/internal/DefaultRequest.java]: Case insensitive would probably be less of a hassle for people.

**Comment by davidmc24** [comment]
Looks good to me.  Thanks again!

**Pull request merged**